File: 프로그래머스/unrated/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        // right : 철수의 토핑종류
        // left  : 동생의 토핑종류
        HashMap<Integer, Integer> right = new HashMap<>();
        Set<Integer> left = new HashSet<>();
        
        // HashMap을 사용해서 전체 토핑종류를 카운팅한다
        for (int i = 0; i < topping.length; i++) {
            right.put(topping[i], right.getOrDefault(topping[i], 0) + 1);
        }
        
        // 다시 for문을 통해 탐색하면서 철수의 토핑종류를 늘려준다
        for (int i = 0; i < topping.length; i++) {
            if (right.get(topping[i]) == 1) {
                right.remove(topping[i]);
            } else {
                right.put(topping[i], right.get(topping[i])-1);
            }
            
            left.add(topping[i]);
            
            // 둘이 토핑의 종류가 같다면 answer의 값을 하나더 올려준다
            if (left.size() == right.size()) {
                answer++;
            }
                          
        }
        return answer;
    }
}

File: 프로그래머스/unrated/132265. 롤케이크 자르기/롤케이크 자르기.py
import collections
def solution(topping):
    answer = 0
    
    # right가 동생의 롤케이크의 토핑 갯수를 count한 hashmap
    # left가  철수의 롤케이크의 토핑 갯수를 count한 hashmap
    right = collections.Counter(topping)
    left = collections.defaultdict(int)
    
    # for문을 통해 왼쪽에서 오른 쪽으로 이동하면서
    # 철수의 토핑갯수가 늘어나고 동생의 토핑 갯수는 줄어든다
    for i, t in enumerate(topping):
        # 동생의 토핑갯수가 1이라면 딕셔너리에서 아예제거한다
        if right[t] == 1:
            right.pop(t)
        else:
            right[t] -=1
        # 철수의 토핑갯수는 늘어난다
        left[t] += 1
        
        # 막약 둘의 토핑갯수가 같다면 경우의 수가 하나씩 늘어나게된다
        if len(right) == len(left):
            answer += 1
        elif len(right) < len(left):
            break
            
    return answer